kimchi/src/circuits/constraints.rs

```diff
@@ -637,7 +616,21 @@ impl<F: FftField + SquareRootField> ConstraintSystem<F> {
             endo,
             fr_sponge_params,
             lookup_constraint_system,
-            precomputations,
+            precomputations: domain_constant_evaluation,
+        };
+
+        constraints.get_or_set_domain_constant_evaluation(precomputations);
+
+        Some(constraints)
+    }
+
+    pub fn get_or_set_domain_constant_evaluation(
+        &self,
+        precomputations: Option<Arc<DomainConstantEvaluations<F>>>,
+    ) -> &Arc<DomainConstantEvaluations<F>> {
+        self.precomputations.get_or_init(|| match precomputations {
+            Some(t) => t,
+            None => Arc::new(DomainConstantEvaluations::create(self.domain).unwrap()),
```

Comment:
not really a fan of this approach for two reasons:

* the function name is long, and you're going to use it in multiple places
* it always takes a `None` everywhere you want to use it

I'd rather have two functions, once that I can call after deserialization, and one that I can call to get the value